warp10/src/main/java/io/warp10/script/functions/FETCH.java

```diff
@@ -235,32 +236,84 @@ public Object apply(WarpScriptStack stack) throws WarpScriptException {
-        } else {
-          throw new WarpScriptException(getName() + " provided token is incompatible with '" + PARAM_GTS + "' parameter, expecting a single owner.");
+        //
+        // If the Metadata have producer/owner/app labels, check if 'matcher' would select them
+        //
+        
+        boolean matches = false;
+        
+        if (m.getLabels().containsKey(Constants.PRODUCER_LABEL)
+            && m.getLabels().containsKey(Constants.OWNER_LABEL)
+            && m.getLabels().containsKey(Constants.APPLICATION_LABEL)) {
+          matches = matcher.matches(m);
         }
         
-        if (m.getLabels().containsKey(Constants.APPLICATION_LABEL) && '=' == m.getLabels().get(Constants.APPLICATION_LABEL).charAt(0)) {
-          m.getLabels().put(Constants.APPLICATION_LABEL, m.getLabels().get(Constants.APPLICATION_LABEL).substring(1));
-        } else {
-          throw new WarpScriptException(getName() + " provided token is incompatible with '" + PARAM_GTS + "' parameter, expecting a single application.");
+        //
+        // If the metadata would not get selected by the provided token
+        // force the producer/owner/app to be that of the token
+        //
+        
+        if (!matches) {
+          m.getLabels().remove(Constants.PRODUCER_LABEL);
+          m.getLabels().remove(Constants.OWNER_LABEL);
+          m.getLabels().remove(Constants.APPLICATION_LABEL);
+              
+          //
+          // If the token doesn't contain a single app we abort the selection as we cannot
+          // choose an app which would be within the reach of the token
+          //
+          
+          if (singleApp) {
+            m.getLabels().put(Constants.APPLICATION_LABEL, tokenSelectors.get(Constants.APPLICATION_LABEL).substring(1));
+          } else {
+            throw new WarpScriptException(getName() + " provided token is incompatible with '" + PARAM_GTS + "' parameter, expecting a single application.");
+          }
+
+          if (singleProducer && singleOwner) {
+            //
+            // If the token has a single producer and single owner, use them for the GTS
+            //
+            m.getLabels().put(Constants.PRODUCER_LABEL, tokenSelectors.get(Constants.PRODUCER_LABEL).substring(1));
+            m.getLabels().put(Constants.OWNER_LABEL, tokenSelectors.get(Constants.OWNER_LABEL).substring(1));            
+          } else if (singleProducer && !tokenSelectors.containsKey(Constants.OWNER_LABEL)) {
+            //
+            // If the token has a single producer but no owner, use the producer as the owner, this would
+            // lead to a narrower scope than what the token would actually select so it is fine.
+            //
+            m.getLabels().put(Constants.PRODUCER_LABEL, tokenSelectors.get(Constants.PRODUCER_LABEL).substring(1));
+            m.getLabels().put(Constants.OWNER_LABEL, m.getLabels().get(Constants.PRODUCER_LABEL));                        
```

Comment:
Using a variable containing `tokenSelectors.get(Constants.PRODUCER_LABEL).substring(1)` would be better IMHO.